### PR TITLE
CBMC: Add spec + proof for `polyvec_basemul_acc_montgomery_cached` + `poly_add`

### PR DESCRIPTION
**Pull request by hanno-becker**
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

### CODE REVIEW

**Comment by hanno-becker** [comment]
@rod-chapman Something strange is happening here: The aliasing-precondition `r==a` in the `poly_add()` contracts seems to be contradictory to CBMC -- you can even put `0==1` in the post-condition, and the proof will go through. _But_, the contract _can_ be used by the caller `polyvec_basemul_acc_montgomery_cached()`. This seems worrying to me. Can you explain what's going on?

**Comment by hanno-becker** [comment]
I used `SAME_OBJECT` + `POINTER_OFFSET` to express pointer equality for now, which seems to work better. Still, the aliasing issue needs investigating.

File: mlkem/poly.h [hunk 1]
```diff
@@ -324,7 +324,36 @@ ENSURES(ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs, 0, MLKEM_Q - 1));
...
+ * NOTE: The restriction r == a is not needed, but we currently only
+ *       prove a CBMC spec under that assumption.
+ *
+ ************************************************************/
+void poly_add(poly *r, const poly *a, const poly *b)
+    // clang-format off
+// So far, poly_add is only used with r == a, so restrict proof to this case
+// TODO: Is this the right way to capture the aliasing constraint a==r?
+REQUIRES(SAME_OBJECT(r,a))
+REQUIRES(__CPROVER_POINTER_OFFSET(r)==__CPROVER_POINTER_OFFSET(a))
```
Comment (mkannwischer) [inline mlkem/poly.h:349]: this should be wrapped as well?

File: mlkem/poly.h [hunk 2]
```diff
@@ -324,7 +324,36 @@ ENSURES(ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs, 0, MLKEM_Q - 1));
...
+ * not overflow. Otherwise, the behaviour of this function is undefined.
+ *
+ * NOTE: The restriction r == a is not needed, but we currently only
+ *       prove a CBMC spec under that assumption.
+ *
+ ************************************************************/
+void poly_add(poly *r, const poly *a, const poly *b)
+    // clang-format off
+// So far, poly_add is only used with r == a, so restrict proof to this case
+// TODO: Is this the right way to capture the aliasing constraint a==r?
```
Comment (mkannwischer) [inline mlkem/poly.h:347]: Can @rod-chapman comment on this, please?

File: mlkem/poly.h [hunk 3]
```diff
@@ -324,7 +324,36 @@ ENSURES(ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs, 0, MLKEM_Q - 1));
...
+REQUIRES(SAME_OBJECT(r,a))
```
Comment (mkannwischer) [inline mlkem/poly.h:348]: Do we additionally need a `r != NULL`?

File: mlkem/polyvec.h [hunk 4]
```diff
@@ -40,9 +40,37 @@ void polyvec_basemul_acc_montgomery(poly *r, const polyvec *a,
...
-                                           const polyvec_mulcache *b_cache);
+                                           const polyvec_mulcache *b_cache)
+    // clang-format off
+REQUIRES(r != NULL && a != NULL && b != NULL && b_cache != NULL)
+REQUIRES(IS_FRESH(r, sizeof(poly)))
+REQUIRES(IS_FRESH(a, sizeof(polyvec)) && IS_FRESH(b, sizeof(polyvec)))
+REQUIRES(IS_FRESH(b_cache, sizeof(polyvec_mulcache)))
+// Input is coefficient-wise < q in absolute value
+REQUIRES(FORALL(int, k1, 0, MLKEM_K - 1,                 \
+ ARRAY_IN_BOUNDS(int, k2, 0, MLKEM_N - 1,                \
```
Comment (mkannwischer) [inline mlkem/polyvec.h:70]: The \ here is not needed, is it?

File: mlkem/polyvec.c [hunk 5]
```diff
@@ -226,18 +226,35 @@ void polyvec_basemul_acc_montgomery_cached(poly *r, const polyvec *a,
...
-    poly_add(r, r, &t);
-    // abs bounds: < (i+1) * 3/2 * q
-  }
 
-  // abs bounds: < MLKEM_K * 3/2 * q <= 4 * 3/2 * q = 19974
+  for (i = 1; i < MLKEM_K; i++)
+      // clang-format off
+    ASSIGNS(i, t, OBJECT_WHOLE(r))
+    INVARIANT(i >= 1 && i <= MLKEM_K)
+    INVARIANT(ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs, \
```
Comment (mkannwischer) [inline mlkem/polyvec.c:238]: The \ here is not needed, is it?

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks @hanno-becker! The pre- and post-conditions are looking good. Some small suggestions. 

Also could you add to the PR title that this also proves poly_add?

File: mlkem/polyvec.c (hunk 5 above)
Comment (hanno-becker) [reply mlkem/polyvec.c:238]: Removing

File: mlkem/polyvec.h (hunk 4 above)
Comment (hanno-becker) [reply mlkem/polyvec.h:70]: Removing

File: mlkem/poly.h (hunk 2 above)
Comment (hanno-becker) [reply mlkem/poly.h:347]: I pinged the CBMC team about this and would file a follow-up PR if the current formulation is not idiomatic. Ok?

File: mlkem/poly.h (hunk 1 above)
Comment (hanno-becker) [reply mlkem/poly.h:349]: Ok

File: mlkem/poly.h (hunk 3 above)
Comment (hanno-becker) [reply mlkem/poly.h:348]: Good point

File: mlkem/poly.h (hunk 2 above)
Comment (hanno-becker) [reply mlkem/poly.h:347]: https://github.com/pq-code-package/mlkem-c-aarch64/issues/323

**Comment by hanno-becker** [comment]
Opened https://github.com/pq-code-package/mlkem-c-aarch64/issues/325

**Comment by hanno-becker** [comment]
@mkannwischer @rod-chapman The reasoning about aliasing in `poly_add()` posed significant verification challenges that are ultimately unnecessary because `poly_add()` is always used in the aliased form.

I have thus specialized `poly_add()` to the accumulator-form, which rendered all proofs straightforward.

**Review by mkannwischer — APPROVED** [review]
Thanks! One typo in a comment - rest looks good to me now.

File: cbmc/proofs/polyvec_basemul_acc_montgomery_cached/polyvec_basemul_acc_montgomery_cached_harness.c [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+/**
+ * @brief Starting point for formal analysis
+ *
+ */
+void harness(void) {
+  poly r;
+  polyvec a, b;
+  polyvec_mulcache b_cached;
```
Comment (mkannwischer) [inline cbmc/proofs/polyvec_basemul_acc_montgomery_cached/polyvec_basemul_acc_montgomery_cached_harness.c:27]: Do you want to change those to pointers already, so you do not have to change them again in #340?

File: mlkem/poly.h [hunk 7]
```diff
@@ -430,7 +430,31 @@ ENSURES(ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs, 0, MLKEM_Q - 1));
...
+ * The coefficients of r and b must be so that the addition does
+ * not overflow. Otherwise, the behaviour of this function is undefined.
+ *
+ ************************************************************/
+// REF-CHANGE:
+// The reference implementation uses a 3-argument poly_add.
+// We specialize to the accumulator form to avoid reasoning about aliasing.
+void poly_add(poly *r, const poly *b)  // clang-format off
+REQUIRES(r != NULL && IS_FRESH(r, sizeof(poly)))
+REQUIRES(b != NULL && IS_FRESH(b, sizeof(poly)))
```
Comment (mkannwischer) [inline mlkem/poly.h:451]: Remove != NULL?

File: mlkem/polyvec.h [hunk 8]
```diff
@@ -40,9 +40,37 @@ void polyvec_basemul_acc_montgomery(poly *r, const polyvec *a,
...
+ *              - const polyvec_mulcache *b_cache: pointer to mulcache
+ *                  for second input polynomial vector. Can be computed
+ *                  via polyvec_mulcache_compute().
+ **************************************************/
 void polyvec_basemul_acc_montgomery_cached(poly *r, const polyvec *a,
                                            const polyvec *b,
-                                           const polyvec_mulcache *b_cache);
+                                           const polyvec_mulcache *b_cache)
+    // clang-format off
+REQUIRES(r != NULL && a != NULL && b != NULL && b_cache != NULL)
```
Comment (mkannwischer) [inline mlkem/polyvec.h:64]: Remove != NULL?

Comment (hanno-becker) [reply mlkem/polyvec.h]: Fixed

File: mlkem/poly.h (hunk 7 above)
Comment (hanno-becker) [reply mlkem/poly.h:451]: Fixed

File: cbmc/proofs/polyvec_basemul_acc_montgomery_cached/polyvec_basemul_acc_montgomery_cached_harness.c (hunk 6 above)
Comment (hanno-becker) [reply cbmc/proofs/polyvec_basemul_acc_montgomery_cached/polyvec_basemul_acc_montgomery_cached_harness.c]: Fixed

File: mlkem/polyvec.c [hunk 9]
```diff
@@ -226,18 +226,35 @@ void polyvec_basemul_acc_montgomery_cached(poly *r, const polyvec *a,
...
+    }
+
+  // Those bounds are true for the C implementation, but not needed
+  // in the higher level bounds reasoning. It is thus best to omit
+  // them from the spec to not unnecessarily constraint native implementations.
+  ASSERT(
+      ARRAY_IN_BOUNDS(int, k, 0, MLKEM_N - 1, r->coeffs,
+                      MLKEM_K * (-3 * HALF_Q + 1), MLKEM_K * (3 * HALF_Q - 1)),
+      "polyvec_basemul_acc_montgomery_cached output bounds");
+  // TODO: Integrate CBMC assertin into POLY_BOUND if CBMC is set
```
[the remark this answers is not in the document]
Comment (hanno-becker) [reply mlkem/polyvec.c]: Fixed

File: mlkem/poly.h (hunk 7 above)
Comment (mkannwischer) [reply mlkem/poly.h:451]: Are you sure? ;-)

**Review by mkannwischer — APPROVED** [review]
Thanks for addressing my comments. Looks good now.

**Pull request merged**